### ISSUE TITLE
Permissions issue where users are locked out of a question they have access to when editing filters

### ISSUE THREAD

**Issue by ipeevski** [metabase/metabase#18433]
**Describe the bug**
Users that should have access to a question, can't access it after they change a filter a second time (the first time works).

**Logs**

```
[bbc10acb-ee33-4499-a187-105cea7a7b1a] 2021-10-13T15:43:16+10:30 ERROR metabase.query-processor.middleware.catch-exceptions Error processing query: null
{:database_id 2,
 :started_at #t "2021-10-13T15:43:16.277413+10:30[Australia/Adelaide]",
 :error_type :missing-required-permissions,
 :json_query
 {:database 2,
  :type "native",
  :native
  {:query
   "select \n    clients.id [[...redacted...]] limit 10",
   :collection "activities",
   :template-tags
   {:name
    {:id "28e10efd-bfaf-d063-593d-50c98c01788f", :name "name", :display-name "Name", :type "text", :required false},
    :email {:id "30c0930b-b463-2237-1159-5282047b6b31", :name "email", :display-name "Email", :type "text"},
    :user {:id "811e536c-b010-123c-aeab-b3c89c37305e", :name "user", :display-name "Mobile Number", :type "number"},
    :patient_id
    {:id "fe0a64ca-a010-c89c-706c-78645dbab927", :name "patient_id", :display-name "Patient", :type "number"}}},
  :parameters [{:type "category", :value "1232345", :target ["variable" ["template-tag" "patient_id"]]}],
  :middleware {:js-int-to-string? true, :add-default-userland-constraints? true}},
 :status :failed,
 :class clojure.lang.ExceptionInfo,
 :stacktrace
 ["--> query_processor.middleware.permissions$perms_exception.invokeStatic(permissions.clj:29)"
  "query_processor.middleware.permissions$perms_exception.doInvoke(permissions.clj:24)"
  "query_processor.middleware.permissions$perms_exception.invokeStatic(permissions.clj:26)"
  "query_processor.middleware.permissions$perms_exception.invoke(permissions.clj:24)"
  "query_processor.middleware.permissions$fn__45548$check_ad_hoc_query_perms__45553$fn__45554.invoke(permissions.clj:83)"
  "query_processor.middleware.permissions$fn__45548$check_ad_hoc_query_perms__45553.invoke(permissions.clj:79)"
  "query_processor.middleware.permissions$fn__45580$check_query_permissions_STAR___45585$fn__45586.invoke(permissions.clj:98)"
  "query_processor.middleware.permissions$fn__45580$check_query_permissions_STAR___45585.invoke(permissions.clj:88)"
  "query_processor.middleware.permissions$check_query_permissions$fn__45599.invoke(permissions.clj:107)"
  "query_processor.middleware.pre_alias_aggregations$pre_alias_aggregations$fn__50109.invoke(pre_alias_aggregations.clj:40)"
  "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__45997.invoke(cumulative_aggregations.clj:60)"
  "query_processor.middleware.visualization_settings$update_viz_settings$fn__45935.invoke(visualization_settings.clj:63)"
  "query_processor.middleware.resolve_joined_fields$resolve_joined_fields$fn__47042.invoke(resolve_joined_fields.clj:102)"
  "query_processor.middleware.resolve_joins$resolve_joins$fn__50695.invoke(resolve_joins.clj:171)"
  "query_processor.middleware.add_implicit_joins$add_implicit_joins$fn__50231.invoke(add_implicit_joins.clj:190)"
  "query_processor.middleware.large_int_id$convert_id_to_string$fn__47061.invoke(large_int_id.clj:59)"
  "query_processor.middleware.format_rows$format_rows$fn__50283.invoke(format_rows.clj:74)"
  "query_processor.middleware.add_default_temporal_unit$add_default_temporal_unit$fn__46433.invoke(add_default_temporal_unit.clj:23)"
  "query_processor.middleware.desugar$desugar$fn__45908.invoke(desugar.clj:21)"
  "query_processor.middleware.binning$update_binning_strategy$fn__39130.invoke(binning.clj:229)"
  "query_processor.middleware.resolve_fields$resolve_fields$fn__45242.invoke(resolve_fields.clj:34)"
  "query_processor.middleware.add_dimension_projections$add_remapping$fn__49483.invoke(add_dimension_projections.clj:312)"
  "query_processor.middleware.add_implicit_clauses$add_implicit_clauses$fn__49781.invoke(add_implicit_clauses.clj:147)"
  "query_processor.middleware.upgrade_field_literals$upgrade_field_literals$fn__47429.invoke(upgrade_field_literals.clj:40)"
  "query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__46810.invoke(add_source_metadata.clj:123)"
  "query_processor.middleware.reconcile_breakout_and_order_by_bucketing$reconcile_breakout_and_order_by_bucketing$fn__50031.invoke(reconcile_breakout_and_order_by_bucketing.clj:100)"
  "query_processor.middleware.auto_bucket_datetimes$auto_bucket_datetimes$fn__48380.invoke(auto_bucket_datetimes.clj:147)"
  "query_processor.middleware.resolve_source_table$resolve_source_tables$fn__45223.invoke(resolve_source_table.clj:45)"
  "query_processor.middleware.parameters$substitute_parameters$fn__48034.invoke(parameters.clj:111)"
  "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__45296.invoke(resolve_referenced.clj:79)"
  "query_processor.middleware.expand_macros$expand_macros$fn__51353.invoke(expand_macros.clj:184)"
  "query_processor.middleware.add_timezone_info$add_timezone_info$fn__47813.invoke(add_timezone_info.clj:15)"
  "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__50298.invoke(splice_params_in_response.clj:32)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__49792$fn__49796.invoke(resolve_database_and_driver.clj:31)"
  "driver$do_with_driver.invokeStatic(driver.clj:60)"
  "driver$do_with_driver.invoke(driver.clj:56)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__49792.invoke(resolve_database_and_driver.clj:25)"
  "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__45846.invoke(fetch_source_query.clj:274)"
  "query_processor.middleware.store$initialize_store$fn__46034$fn__46035.invoke(store.clj:11)"
  "query_processor.store$do_with_store.invokeStatic(store.clj:44)"
  "query_processor.store$do_with_store.invoke(store.clj:38)"
  "query_processor.middleware.store$initialize_store$fn__46034.invoke(store.clj:10)"
  "query_processor.middleware.validate$validate_query$fn__50081.invoke(validate.clj:10)"
  "query_processor.middleware.normalize_query$normalize$fn__50088.invoke(normalize_query.clj:22)"
  "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__47759.invoke(add_rows_truncated.clj:35)"
  "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__49057.invoke(results_metadata.clj:147)"
  "query_processor.middleware.constraints$add_default_userland_constraints$fn__47777.invoke(constraints.clj:42)"
  "query_processor.middleware.process_userland_query$process_userland_query$fn__50067.invoke(process_userland_query.clj:146)"
  "query_processor.middleware.catch_exceptions$catch_exceptions$fn__50378.invoke(catch_exceptions.clj:169)"
  "query_processor.reducible$async_qp$qp_STAR___42481$thunk__42482.invoke(reducible.clj:103)"
  "query_processor.reducible$async_qp$qp_STAR___42481.invoke(reducible.clj:109)"
  "query_processor.reducible$sync_qp$qp_STAR___42490$fn__42493.invoke(reducible.clj:135)"
  "query_processor.reducible$sync_qp$qp_STAR___42490.invoke(reducible.clj:134)"
  "query_processor$process_userland_query.invokeStatic(query_processor.clj:241)"
  "query_processor$process_userland_query.doInvoke(query_processor.clj:237)"
  "query_processor$fn__51401$process_query_and_save_execution_BANG___51410$fn__51413.invoke(query_processor.clj:253)"
  "query_processor$fn__51401$process_query_and_save_execution_BANG___51410.invoke(query_processor.clj:245)"
  "query_processor$fn__51445$process_query_and_save_with_max_results_constraints_BANG___51454$fn__51457.invoke(query_processor.clj:265)"
  "query_processor$fn__51445$process_query_and_save_with_max_results_constraints_BANG___51454.invoke(query_processor.clj:258)"
  "api.dataset$run_query_async$fn__57767.invoke(dataset.clj:63)"
  "query_processor.streaming$streaming_response_STAR_$fn__38039$fn__38040.invoke(streaming.clj:160)"
  "query_processor.streaming$streaming_response_STAR_$fn__38039.invoke(streaming.clj:159)"
  "async.streaming_response$do_f_STAR_.invokeStatic(streaming_response.clj:65)"
  "async.streaming_response$do_f_STAR_.invoke(streaming_response.clj:63)"
  "async.streaming_response$do_f_async$fn__25228.invoke(streaming_response.clj:84)"],
 :card_id nil,
 :context :ad-hoc,
 :error "You do not have permissions to run this query.",
 :row_count 0,
 :running_time 0,
 :ex-data
 {:type :missing-required-permissions,
  :required-permissions #{"/db/2/native/"},
  :actual-permissions
  #{"/collection/291/"
    "/db/10/schema/"
    "/collection/258/"
    "/collection/84/"
    "/collection/282/"
    "/collection/249/read/"
    "/collection/279/"
    "/collection/238/"
    "/collection/74/read/"
    "/collection/39/"
    "/collection/244/"
    "/collection/284/"
    "/collection/259/"
    "/collection/1/"
    "/collection/240/"
    "/collection/237/"
    "/collection/285/"
    "/collection/268/"
    "/collection/267/"
    "/collection/255/"
    "/collection/251/"
    "/collection/272/"
    "/collection/3/"
    "/collection/290/"
    "/collection/253/"
    "/collection/262/"
    "/collection/292/"
    "/collection/248/read/"
    "/collection/293/"
    "/db/2/schema/"
    "/collection/241/"
    "/collection/269/"
    "/collection/295/"
    "/collection/261/read/"
    "/collection/273/"
    "/collection/287/"
    "/db/13/"
    "/collection/294/"
    "/collection/270/"
    "/collection/7/"
    "/collection/276/"
    "/collection/109/"
    "/collection/76/"
    "/collection/141/"
    "/collection/266/"
    "/collection/275/"
    "/collection/264/"
    "/collection/265/"
    "/collection/278/"
    "/collection/242/"
    "/collection/42/"
    "/collection/263/"
    "/collection/256/read/"
    "/collection/239/"
    "/collection/252/"
    "/collection/40/"
    "/collection/289/"
    "/collection/288/"
    "/collection/280/"
    "/collection/274/"},
  :permissions-error? true},
 :data {:rows [], :cols []}}
```

From the logs, I can see that the query was ran as a native query, even though it was accessed from a question, without any modifications other than changing a filter.

**To Reproduce**
1. Create a question with a filter in a collection
2. Create a user and give them access to only that collection
3. Go to the question as that user, fill in the filter and press search
4. Edit the filter and press search again
5. Get an error "You do not have permissions to run this query."
6. The URL at the top after the second search has a hash (which seems to indicate that it ran a modified raw query). The URL after the first search doesn't have a hash, just a `?param` added at the end.

**Expected behavior**
Both searches work correctly as the user has access to the question.
This is a regression, as it used to work fine in 0.38 (maybe even after that).

**Information about your Metabase Installation:**
- Your browser and the version: Chrome 94
- Your operating system: Ubuntu 21.04
- Your databases: MySQL
- Metabase version: 0.41.0
- Metabase hosting environment: Docker
- Metabase internal database: MySQL

```
{
  "browser-info": {
    "language": "en",
    "platform": "Linux x86_64",
    "userAgent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.81 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.12+7",
    "java.vendor": "Eclipse Foundation",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "11.0.12",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.12+7",
    "os.name": "Linux",
    "os.version": "4.14.239-aptible.0",
    "user.language": "en",
    "user.timezone": "Australia/Adelaide"
  },
  "metabase-info": {
    "databases": [
      "mysql",
      "googleanalytics"
    ],
    "hosting-env": "unknown",
    "application-database": "mysql",
    "application-database-details": {
      "database": {
        "name": "MySQL",
        "version": "8.0.25"
      },
      "jdbc-driver": {
        "name": "MariaDB Connector/J",
        "version": "2.6.2"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2021-10-06",
      "tag": "v0.41.0",
      "branch": "release-x.41.x",
      "hash": "c529fe2"
    },
    "settings": {
      "report-timezone": "Australia/Adelaide"
    }
  }
}
```

**Severity**
There is a work around - you can go back to the question and do a new search and that works, but it does get annoying and might force us to downgrade to mitigate the issue.

**Additional context**
It's a regression - it used to work fine in 0.38

**Comment by ipeevski** [comment]
Hi @flamber 
I couldn't reproduce it with a new question either now.
And while trying, I did find a work around. After I do the second search (as admin) and it populates the hash in the URL. If I save the question again (overwrite the old one), then it starts working as expected.
So something in how the old question was saved vs how it stores it in the latest version has changed is my guess.

To replicate the issue, I'm guessing you'll have to install 0.38, create the question with the filter, then upgrade.
I couldn't get steps to replicate it with sample database.

For reference, **workaround**:
- Save the question again

**Comment by araujo-filipe** [comment]
We got to this issue a little bit differently than described but seems to be exactly the same problem.
1. Create a question with a text filter on a collection
2. Create a user and give them access to only that collection
3. As that user, add a large value to the filter - in our case, more than 4700 characters did the trick
4. As that user, run the question (The url now is similar to ad-hoc question urls)
5. As that user, try to download the result or change the filter again and hit the permissions error

I tried re-saving the question, but that did not fix the issue...
Any other possible workarounds to make this work while no fix is deployed?
